crates/bevy_scene/src/dynamic_scene_builder.rs

```diff
@@ -0,0 +1,230 @@
+use crate::{DynamicEntity, DynamicScene};
+use bevy_app::AppTypeRegistry;
+use bevy_ecs::{prelude::Entity, reflect::ReflectComponent, world::World};
+use bevy_utils::{default, HashMap};
+
+/// A [`DynamicScene`] builder, used to build a scene from a [`World`] by extracting some entities.
+///
+/// ```
+/// # use bevy_scene::DynamicSceneBuilder;
+/// # use bevy_app::AppTypeRegistry;
+/// # use bevy_ecs::{
+/// #     component::Component, prelude::Entity, query::With, reflect::ReflectComponent, world::World,
+/// # };
+/// # use bevy_reflect::Reflect;
+/// # #[derive(Component, Reflect, Default, Eq, PartialEq, Debug)]
+/// # #[reflect(Component)]
+/// # struct ComponentA;
+/// # let mut world = World::default();
+/// # world.init_resource::<AppTypeRegistry>();
+/// # let entity = world.spawn(ComponentA).id();
+/// let mut builder = DynamicSceneBuilder::from_world(&world);
+/// builder.extract_entity(entity);
+/// let dynamic_scene = builder.build();
+/// ```
+pub struct DynamicSceneBuilder<'w> {
+    scene: HashMap<u32, DynamicEntity>,
+    type_registry: AppTypeRegistry,
+    world: &'w World,
+}
+
+impl<'w> DynamicSceneBuilder<'w> {
+    /// Prepare a builder that will extract entities and their component from the given [`World`].
+    /// All components registered in that world's [`AppTypeRegistry`] resource will be extracted.
+    pub fn from_world(world: &'w World) -> Self {
+        Self {
+            scene: default(),
+            type_registry: world.resource::<AppTypeRegistry>().clone(),
+            world,
+        }
+    }
+
+    /// Prepare a builder that will extract entities and their component from the given [`World`].
+    /// Only components registered in the given [`AppTypeRegistry`] will be extracted.
+    pub fn from_world_with_type_registry(world: &'w World, type_registry: AppTypeRegistry) -> Self {
```

Comment:
Nit: This name is just a bit wordy imo. Since this is a builder struct, I wonder if we could turn this into:
```rust
pub fn with_registry(self, type_registry: AppTypeRegistry) -> Self {
  Self { type_registry, ..self }
}
```

Although, this might allow some behavior I don't know if we'd want (maybe we do), such as swapping out the registry between extractions:

```rust
let mut builder = DynamicSceneBuilder::from_world(&world);
builder.extract_entity(entity_a);
let mut builder = builder.with_registry(&type_filter);
builder.extract_entity(entity_b);
```

Reply:
changing the registry between extractions could complicate things. it would mean the behaviour when extracting twice an entity with different registries is not obvious. unless this is a feature that is actually needed, I would rather avoid it

Reply:
Yeah, I agree. I think that's likely an anti-feature.

Reply:
Makes sense, we can exclude this suggestion for now!